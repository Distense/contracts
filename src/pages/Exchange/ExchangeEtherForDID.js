import React, { Component } from 'react'
import { BigNumber } from 'bignumber.js'
import { Button, Input, Form, Grid, List, Message } from 'semantic-ui-react'

import { investEtherForDID } from '../../actions/exchange'

export default class ExchangeEtherForDID extends Component {
  constructor(props) {
    super(props)
    this.state = {
      numEtherUserWantsToExchange: '',
      didPerEtherExchangeRate: this.props.didPerEtherExchangeRate || 0
    }
    this.onChangeNumEther = this.onChangeNumEther.bind(this)
  }

  componentDidMount() {
    const numDIDOwned = this.props.numDIDOwned
    this.setState({
      numDIDOwnedByUser: numDIDOwned,
      didPerEtherExchangeRate: this.props.didPerEtherExchangeRate
    })
  }

  onChangeNumEther = ({ target: { value } }) => {
    this.setState({ numEtherUserWantsToExchange: value })
  }

  onSubmitInvestEtherForDID = async e => {
    e.preventDefault()

    const { numEtherUserWantsToExchange } = this.state

    investEtherForDID({ numEtherUserWantsToExchange })
  }

  render() {
    const { numEtherToExchange } = this.state

    const { numDIDOwned } = this.props

    return (
      <Grid.Column width={8}>
        <Form onSubmit={this.onSubmitInvestEtherForDID}>
          <Form.Field>
            <Input
              type="text"
              placeholder="receive DID, give up ether"
              onChange={this.onChangeNumEther}
              className=""
              name="title"
              value={numEtherToExchange}
            />
          </Form.Field>
          <Button basic size="large" color="green" type="submit">
            Invest ether for DID
          </Button>
        </Form>
        <Message>
          <Message.Header>Invest Ether for DID</Message.Header>
          <List bulleted>
            <List.Item>
              You may exchange ether only to the extent you have contributed to
              Distense. I.e, according to the number of DID you{' '}
              <em>continue</em> to own
            </List.Item>
            <List.Item>
              DID have a very high chance of being worthless. DID are not traded
              on any exchange and are non-transferable. Lots has to happen for
              DID to be worth anything.
            </List.Item>
            <List.Item>
              If you are idealistically aligned with Distense, have a super
              long-term timeframe in mind, then maybe, just maybe, you should
              consider giving up your precious ether for DID.
            </List.Item>
            <List.Item>
              This is a permanent, non-reversible, undoable, final action. There
              is no one who can undo this. Ever.
            </List.Item>
          </List>
        </Message>
      </Grid.Column>
    )
  }
}
