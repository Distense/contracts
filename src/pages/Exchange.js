import React, { Component } from 'react'

import { Grid, List, Message } from 'semantic-ui-react'

import ExchangeDIDForEther from './Exchange/ExchangeDIDForEther'
import ExchangeEtherForDID from './Exchange/ExchangeEtherForDID'

import Head from '../components/common/Head'
import PageTitling from '../components/common/PageTitling'

export default class Exchange extends Component {
  render() {
    return (
      <div>
        <Head title="Exchange DID for ETH" />
        <Grid>
          <PageTitling
            title="Exchange"
            subtitle="DID holders may exchange their DID into ether or vice versa"
          />
          <Grid.Column width={8}>
            <Message>
              <Message.Header>Distense Finances</Message.Header>
              <List bulleted>
                <List.Item>
                  ETH in Distense bank account: {this.props.numBankAccountEther}
                </List.Item>
                <List.Item>
                  Exchange rate: {this.props.didPerEtherExchangeRate} DID per
                  ETH
                </List.Item>
              </List>
            </Message>
          </Grid.Column>

          <Grid.Row style={{ paddingTop: '0px' }} columns={2}>
            <ExchangeDIDForEther {...this.props} />
            <ExchangeEtherForDID {...this.props} />
          </Grid.Row>
          <Grid.Row />
        </Grid>
      </div>
    )
  }
}
