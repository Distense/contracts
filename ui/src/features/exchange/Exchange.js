import React, { Component } from 'react'

import { Grid, List, Message } from 'semantic-ui-react'

import ExchangeDIDForEther from './components/ExchangeDIDForEther'
import ExchangeEtherForDID from './components/ExchangeEtherForDID'

import Head from '../../components/Head'
import PageTitling from '../../components/PageTitling'

export default class Exchange extends Component {
  render() {
    return (
      <div>
        <Head title="Exchange DID for ETH and vice versa" />
        <Grid>
          <PageTitling
            title="Exchange"
            subtitle="DID holders may exchange their DID into ether or vice versa"
          />
          <Grid.Column width={16}>
            <Message>
              <Message.Header>Distense Finances</Message.Header>
              <List bulleted>
                <List.Item>
                  ETH in Distense bank account: {this.props.numBankAccountEther}
                </List.Item>
                <List.Item>
                  Exchange rate: {this.props.didPerEtherExchangeRate} DID per
                  ETH
                </List.Item>
              </List>
            </Message>
          </Grid.Column>

          <Grid.Row style={{ paddingTop: '0px' }} columns={2}>
            <ExchangeDIDForEther {...this.props} />
            <ExchangeEtherForDID {...this.props} />
          </Grid.Row>
          <Grid.Row />
        </Grid>
      </div>
    )
  }
}
