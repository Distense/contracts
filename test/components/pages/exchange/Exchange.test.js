import React from 'react'
import { shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import Exchange from '../../../../src/features/Exchange'

function setup(numBankAccountEther = null, didPerEtherExchangeRate = null) {
  const props = {
    numBankAccountEther,
    didPerEtherExchangeRate
  }
  let wrapper = shallow(<Exchange {...props} />)

  return { wrapper, props }
}

describe('<Exchange /> page component', () => {
  it('should contain some basic subcomponents', () => {
    const { wrapper } = setup(123, 1000)

    expect(wrapper.find('Grid').length).toEqual(1)
    expect(wrapper.find('GridColumn').length).toEqual(1)
    expect(wrapper.find('Message').length).toEqual(1)
    expect(wrapper.find('MessageHeader').length).toEqual(1)
    expect(wrapper.find('List').length).toEqual(1)
    expect(wrapper.find('GridRow').length).toEqual(2)
    expect(wrapper.find('ExchangeDIDForEther').length).toEqual(1)
    expect(wrapper.find('ExchangeEtherForDID').length).toEqual(1)
  })

  test('snapshot', () => {
    const { props } = setup('executive')
    const mockstore = configureMockStore()
    const store = mockstore({})
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Exchange {...props} />
          </MemoryRouter>
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
