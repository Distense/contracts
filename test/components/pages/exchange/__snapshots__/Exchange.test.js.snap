// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exchange /> page component snapshot 1`] = `
<div>
  <div
    className="ui grid"
  >
    <div
      className="sixteen wide column"
    >
      <h2
        className="ui header"
      >
        Exchange
      </h2>
      <p
        style={
          Object {
            "fontSize": "1.3rem",
          }
        }
      >
        DID holders may exchange their DID into ether or vice versa
      </p>
    </div>
    <div
      className="eight wide column"
    >
      <div
        className="ui message"
      >
        <div
          className="header"
        >
          Distense Finances
        </div>
        <div
          className="ui bulleted list"
          role="list"
        >
          <div
            className="item"
            data-value={undefined}
            onClick={[Function]}
            role="listitem"
          >
            ETH in Distense bank account: 
            executive
          </div>
          <div
            className="item"
            data-value={undefined}
            onClick={[Function]}
            role="listitem"
          >
            Exchange rate: 
             DID per ETH
          </div>
        </div>
      </div>
    </div>
    <div
      className="two column row"
      style={
        Object {
          "paddingTop": "0px",
        }
      }
    >
      <div
        className="eight wide column"
      >
        <form
          action={undefined}
          className="ui form"
          onSubmit={[Function]}
        >
          <div
            className="field"
          >
            <div
              className="ui input"
            >
              <input
                disabled={undefined}
                onChange={[Function]}
                placeholder="number of DID you want to exchange into ether"
                tabIndex={undefined}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="ui green medium button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
            type="submit"
          >
            Exchange 
            
             DID for 
            
             ether
          </button>
        </form>
        <div
          className="ui message"
        >
          <div
            className="header"
          >
            Exchange DID For ETH
          </div>
          <div
            className="ui bulleted list"
            role="list"
          >
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You don't own any DID in the selected Ethereum account so you have no DID to exchange
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You are relinquishing Distense governance/voting rights by exchanging your DID for ether.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              This is a permanent, non-reversible, undoable, final action. There is no one who can undo this.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              If this says you can't invest, try reloading the page or ensure MetaMask is unlocked
            </div>
          </div>
        </div>
      </div>
      <div
        className="eight wide column"
      >
        <form
          action={undefined}
          className="ui form"
          onSubmit={[Function]}
        >
          <div
            className="field"
          >
            <div
              className="ui input"
            >
              <input
                disabled={undefined}
                name="title"
                onChange={[Function]}
                placeholder="number of ETH you want to invest in exchange for DID"
                tabIndex={undefined}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="ui green medium button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
            type="submit"
          >
            Invest 
            
             ether for 
             DID
          </button>
        </form>
        <div
          className="ui message"
        >
          <div
            className="header"
          >
            Invest Ether for DID
          </div>
          <div
            className="ui bulleted list"
            role="list"
          >
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You don't own any DID in the selected Ethereum account so you have no DID to exchange
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              You may exchange ether only to the extent you have contributed to Distense. I.e, according to the number of DID you
               
              <em>
                continue
              </em>
               to own
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              DID have a very high chance of being worthless. DID are not traded on any exchange and are non-transferable. Lots has to happen for DID to be worth anything.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              If you are idealistically aligned with Distense, have a super long-term timeframe in mind, then maybe, just maybe, you should consider giving up your precious ether for DID.
            </div>
            <div
              className="item"
              data-value={undefined}
              onClick={[Function]}
              role="listitem"
            >
              This is a permanent, non-reversible, undoable, final action. There is no one who can undo this. Ever.
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    />
  </div>
</div>
`;
